[tune] Distributed example + walkthrough (#5157)
ray-project/ray: doc/source/tune-distributed.rst

Distributed Experiments
=======================

Tune is commonly used for large-scale distributed hyperparameter optimization. Tune provides many utilities that enable an effective workflow for interacting with a cluster.

> To launch a large hyperparameter search, you can download this `configuration file <https://raw.githubusercontent.com/ray-project/ray/master/python/ray/autoscaler/aws/example-full.yaml>`__ and `an example Tune script <https://raw.githubusercontent.com/ray-project/ray/master/python/ray/tune/examples/async_hyperband_example.py>`__ and then run:

Reviewer: I like the way things are listed in the Tune Example Walkthrough below (https://github.com/ray-project/ray/pull/5157/files#diff-f33a6d900fda4d491912af633fe71142R1). Could do something similar here.

Running on spot instances (or pre-emptible instances) can reduce the cost of your experiment. You can enable spot instances in AWS via the following configuration modification:

.. code-block:: yaml

    # Provider-specific config for worker nodes, e.g. instance type.
    worker_nodes:
        InstanceType: m5.large
        ImageId: ami-0b294f219d14e6a82 # Deep Learning AMI (Ubuntu) Version 21.0

        # Run workers on spot by default. Comment this out to use on-demand.
        InstanceMarketOptions:
            MarketType: spot
            SpotOptions:
                MaxPrice: 1.0  # Max Hourly Price

In GCP, you can use the following configuration modification:

.. code-block:: yaml

    worker_nodes:
        machineType: n1-standard-2
        disks:
          - boot: true
            autoDelete: true
            type: PERSISTENT
            initializeParams:
              diskSizeGb: 50
              # See https://cloud.google.com/compute/docs/images for more images
              sourceImage: projects/deeplearning-platform-release/global/images/family/tf-1-13-cpu

        # Run workers on preemtible instances.
        scheduling:
          - preemptible: true

> Spot instances may be removed suddenly while trials are still running. Often times this may be difficult to deal with when using other distributed hyperparameter optimization frameworks. Tune allows users to mitigate the effects of this by preserving the progress of your model training through checkpointing. The easiest way to do this is to subclass the pre-defined ``Trainable`` class and implement ``_save``, and ``_restore`` abstract methods, as seen in `this example <https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/hyperband_example.py>`__. See the `Checkpointing <tune-usage.html#trainable-trial-checkpointing>`__ section for more details.

Reviewer 2: Yeah, that would be a great idea. Something that worked for me was:
1. start a ray cluster up and start a job with tune
2. from the GCP console kill an instance
3. observe in the terminal opened in stage 1

I think it's quite important to put emphasis on this feature, and walk people through the steps needed to get a working config file that's ready for this. I personally would have jumped on the ray train a lot sooner if this feature was more obvious for me.

Tune Distributed Experiments
============================

> Tune is commonly used for large-scale distributed hyperparameter optimization. Tune and Ray provides many utilities that enable an effective workflow for interacting with a cluster, including fast uploading, one-line cluster launching, and uploading results.

Author: ```suggestion
Tune is commonly used for large-scale distributed hyperparameter optimization. Tune and Ray provide many utilities that enable an effective workflow for interacting with a cluster, including fast uploading, one-line cluster launching, and uploading results.
```

> This page will overview how to start a distributed experiment.

Author: ```suggestion
This page will overview the tooling for distributed experiments, covering how to connect to a cluster, how to launch a distributed experiment, and commonly used commands.
```

Connecting to a cluster
-----------------------

One common approach to modifying an existing Tune experiment to go distributed is to set an `argparse` variable so that toggling between distributed and single-node is seamless. This allows Tune to utilize all the resources available to the Ray cluster.

.. code-block:: python

    import ray
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("--ray-redis-address")
    args = parser.parse_args()
    ray.init(redis_address=args.ray_redis_address)

> After, compare single node vs cluster execution. Note that connecting to cluster requires a pre-existing Ray cluster to be started already (`Manual Cluster Setup <using-ray-on-a-cluster.html>`_). The script should be run on the head node of the Ray cluster. Below, ``tune_script.py`` can be any script that runs a Tune hyperparameter search.

Author: ```suggestion
Note that connecting to cluster requires a pre-existing Ray cluster to be started already (`Manual Cluster Setup <using-ray-on-a-cluster.html>`_). The script should be run on the head node of the Ray cluster. Below, ``tune_script.py`` can be any script that runs a Tune hyperparameter search.
```

              sourceImage: projects/deeplearning-platform-release/global/images/family/tf-1-13-cpu

        # Run workers on preemtible instances.
        scheduling:
          - preemptible: true

Spot instances may be removed suddenly while trials are still running. Often times this may be difficult to deal with when using other distributed hyperparameter optimization frameworks. Tune allows users to mitigate the effects of this by preserving the progress of your model training through checkpointing.

The easiest way to do this is to subclass the pre-defined ``Trainable`` class and implement ``_save``, and ``_restore`` abstract methods, as seen in the example below:

.. literalinclude:: ../../python/ray/tune/tests/tutorial.py
   :language: python
   :start-after: __trainable_example_begin__
   :end-before: __trainable_example_end__

This can then be used similarly to the Function API as before: ``tune.run(TrainMNIST, ...)``.

Example for using spot instances (AWS)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here is an example for running Tune on spot instances. This assumes your AWS credentials have already been setup (``aws configure``):

> 1. Download a full example Tune experiment script here: :download:`mnist_pytorch_trainable.py <../../python/ray/tune/examples/mnist_pytorch_trainable.py>`

Author: ```suggestion
1. Download a full example Tune experiment script here. This includes a Trainable with checkpointing: :download:`mnist_pytorch_trainable.py <../../python/ray/tune/examples/mnist_pytorch_trainable.py>`
```

How To: Distributed Tuning
==========================

> This is an example walkthrough for a distributed hyperparameter search in Tune. At a high level, we will use Ray's cluster launcher/autoscaler utility to start a cluster of nodes, all of which are connected by Ray. Then, we will modify an existing hyperparameter tuning script to connect to the Ray cluster, and launch the script. Finally, we will analyze the results.

Reviewer 3: Remove comma before "and launch the script"

Reviewer 3: I think this would be more readable as a list, e.g.:

This is an example walkthrough for distributed hyperparameter search in Tune, consisting of the following steps:
1. Start a Ray cluster of autoscaling nodes using built-in Ray utilities.
2. Modify and launch an existing hyperparameter tuning script to connect to the Ray cluster.
3. Analyze the results.

Quick Start
-----------

.. code-block::bash

>     export CLUSTER=[path/to/cluster/yaml]

Reviewer 3: Not exactly sure why but I found this hard to read. Would probably prefer either `[path_to_cluster_yaml]` or `cluster_config.yaml`.
